### PR TITLE
[WIP] Use BlasLtMatmul APIs in matmul_op_impl

### PR DESCRIPTION
**Pull request by benbarsdell**
This is a rebase of https://github.com/tensorflow/tensorflow/pull/43237 (part of which was reverted due to regressions) to be used for further performance investigation. I've labelled it WIP because we should identify the cause of the regressions before merging.

- Integrates BlasLtMatmul with autotuning into the implementation of the BatchMatMul and Einsum ops.
- This integration is only used when the CUDA version is >= 11.0.

cc @nluehr @timshen91

### CODE REVIEW

File: tensorflow/core/kernels/matmul_op_impl.h [hunk 1]
```diff
@@ -357,31 +531,164 @@ struct LaunchBatchMatMul<GPUDevice, Scalar> {
...
+
+        se::blas::BlasLtMatmulPlanParams plan_params;
+        plan_params.ab_type = blas_dtype;
+        plan_params.c_type = blas_dtype;
+        plan_params.computation_type = computation_type;
+        plan_params.pointer_mode = se::blas::PointerMode::kHost;
+        plan_params.epilogue = se::blas::Epilogue::kDefault;
+        plan_params.transa = blas_transpose_b;
+        plan_params.transb = blas_transpose_a;
+        plan_params.m = n;
```
Comment (AyanmoI) [inline tensorflow/core/kernels/matmul_op_impl.h:587]: Why is `transa`, transb`, `m` and `n` flipped here? Essentially it seems they are flipped when creating matmul_paramters. I don't understand why that needs to be that way. `m` is essentially the num_rows and `n` the num columns of the output.

Comment (AyanmoI) [reply tensorflow/core/kernels/matmul_op_impl.h:587]: XLA doesn't see any conjugate transpose. IIUC can we just use `plan_params.m = m;` etc. where `int64 m = output_matrix.num_cols;` in XLA thunk.

Comment (AyanmoI) [reply tensorflow/core/kernels/matmul_op_impl.h:587]: Also, it seems `plan_params.lda = lhs_matrix.num_rows;` ,`plan_params.ldb = rhs_matrix.num_rows;` and `plan_params.ldc = output_matrix.num_rows;`

**Comment by gbaned** [comment]
@benbarsdell Can you please check @AyanmoI's comments and resolve conflicts?. Thanks!

**Comment by gbaned** [comment]
@benbarsdell  Any update on this PR? Please. Thanks!

File: tensorflow/core/kernels/matmul_op_impl.h (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/core/kernels/matmul_op_impl.h:587]: > @benbarsdell Any update on this PR? Please. Thanks!

@gbaned @benbarsdell has clarified my doubts offline. Thanks for following up though.

**Comment by gbaned** [comment]
@benbarsdell Any update on this PR? and please resolve conflicts Thanks!

**Comment by tensorflowbutler** [comment]
It has been 19 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by benbarsdell** [comment]
Closing for now.

**Pull request closed without merging**